metaflow/plugins/argo/argo_workflows.py

```diff
@@ -290,18 +286,99 @@ def trigger(cls, name, parameters=None):
         except Exception as e:
             raise ArgoWorkflowsException(str(e))
 
-    @staticmethod
-    def _get_kubernetes_labels():
+    def _get_kubernetes_labels(self):
         """
-        Get Kubernetes labels from environment variable.
-        Parses the string into a dict and validates that values adhere to Kubernetes restrictions.
+        Get Kubernetes labels from the start step decorator.
         """
-        if not KUBERNETES_LABELS:
-            return {}
-        env_labels = KUBERNETES_LABELS.split(",")
-        env_labels = parse_kube_keyvalue_list(env_labels, False)
-        validate_kube_labels(env_labels)
-        return env_labels
+
+        resources = dict(
+            [
+                deco
+                for node in self.graph
+                if node.name == "start"
+                for deco in node.decorators
+                if deco.name == "kubernetes"
+            ][0].attributes
+        )
+        return resources["labels"] or {}
+
+    def _get_kubernetes_annotations(self):
+        """
+        Get Kubernetes annotations from the start step decorator.
+        Append Argo specific annotations that are unavailable at the decorator level.
+        """
+        resources = dict(
+            [
+                deco
+                for node in self.graph
+                if node.name == "start"
+                for deco in node.decorators
+                if deco.name == "kubernetes"
+            ][0].attributes
+        )
+        annotations = {}
+        if resources["annotations"] is not None:
+            # make a copy so we do not mess possible start-step specific annotations.
+            annotations = resources["annotations"].copy()
+
+        annotations.update(
+            {
+                "metaflow/production_token": self.production_token,
+                "metaflow/owner": self.username,
+                "metaflow/user": "argo-workflows",
+                "metaflow/flow_name": self.flow.name,
+            }
+        )
+
+        if self.parameters:
+            annotations.update({"metaflow/parameters": json.dumps(self.parameters)})
+
```

Comment:
Moved the new convenience annotations to this helper as well. One thing to note though is that this applies all of them to the generic `self.kubernetes_annotations`, which is applied generously as a baseline set of annotations (including for Sensors)

Do these annotations make sense for all objects, or should they be kept out of this helper? This simplifies the setup somewhat at least.

Reply:
@savingoyal Do you have an opinion on this?